signer/core/signed_data.go

```
@@ -0,0 +1,788 @@
+		Value: typedData.formatData(typedData.PrimaryType, typedData.Message),
+		Typ:   "primary type",
+	})
+	return nvts
+}
+
+func (typedData *TypedData) formatData(primaryType string, data map[string]interface{}) []*NameValueType {
+	var output []*NameValueType
+
+	// Add field contents. Structs and arrays have special handlers.
+	for _, field := range typedData.Types[primaryType] {
+		encName := field.Name
+		encValue := data[encName]
+		item := &NameValueType{
+			Name: encName,
+			Typ:  field.Type,
+		}
+		if field.isArray() {
+			arrayValue, _ := encValue.([]interface{})
+			parsedType := field.typeName()
+			for _, v := range arrayValue {
+				if typedData.Types[parsedType] != nil {
+					mapValue, _ := v.(map[string]interface{})
+					mapOutput := typedData.formatData(parsedType, mapValue)
+					item.Value = mapOutput
+				} else {
+					primitiveOutput := formatPrimitiveValue(field.Type, encValue)
+					item.Value = primitiveOutput
+				}
+			}
+		} else if typedData.Types[field.Type] != nil {
+			mapValue, _ := encValue.(map[string]interface{})
+			mapOutput := typedData.formatData(field.Type, mapValue)
+			item.Value = mapOutput
+		} else {
+			primitiveOutput := formatPrimitiveValue(field.Type, encValue)
+			item.Value = primitiveOutput
+		}
+		output = append(output, item)
+	}
+	return output
+}
+
+func formatPrimitiveValue(encType string, encValue interface{}) string {
+	switch encType {
+	case "address":
+		stringValue, _ := encValue.(string)
+		return common.HexToAddress(stringValue).String()
+	case "bool":
+		boolValue, _ := encValue.(bool)
+		return fmt.Sprintf("%t", boolValue)
+	case "bytes", "string":
+		return fmt.Sprintf("%s", encValue)
+	}
+	if strings.HasPrefix(encType, "bytes") {
+		return fmt.Sprintf("%s", encValue)
+	} else if strings.HasPrefix(encType, "uint") || strings.HasPrefix(encType, "int") {
+		bigIntValue, _ := encValue.(*big.Int)
+		return fmt.Sprintf("%d (0x%x)", bigIntValue, bigIntValue)
+	}
+	return "NA"
```

let's return an error here: when it's called by `SignTypedData` the user will want to know why their request is not accepted.